File: framework/src/main/java/com/example/framework/view/ProgressView.java
package com.example.framework.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.framework.R;

public class ProgressView extends View {
    private Paint paint;
    private final int START_ANGLE = 0;//以3点钟方向开始绘制扇形
    private final int STEP_ANGLE = 1;//每次按照1度的大小进行增加
    private int offsetAngle;
    private int progressAngle = 0;
    private int progressViewWidth;
    private int progressViewHeight;
    private int progress;
    private final int CIRCLE_MARGIN = 17;
    private int textColor;
    private int textSize;
    private int paintSize;
    private int paintColor1;
    private int paintColor2;

    public ProgressView(Context context) {
        this(context, null);
    }

    public ProgressView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ProgressView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        this.paint = new Paint();

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.ProgressView);

        textColor = typedArray.getColor(R.styleable.ProgressView_TextColor, Color.BLACK);
        textSize = typedArray.getInt(R.styleable.ProgressView_TextSize, 30);
        paintSize = typedArray.getInt(R.styleable.ProgressView_PaintSize, 5);
        paintColor1 = typedArray.getInt(R.styleable.ProgressView_PaintColor1, Color.BLUE);
        paintColor2 = typedArray.getInt(R.styleable.ProgressView_PaintColor2, Color.RED);

        typedArray.recycle();
    }

    public void setTextColor(int textColor) {
        this.textColor = textColor;
        invalidate();
    }

    public void setTextSize(int textSize) {
        this.textSize = textSize;
        invalidate();
    }

    public void setPaintSize(int paintSize) {
        this.paintSize = paintSize;
        invalidate();
    }

    public void setPaintColor1(int paintColor1) {
        this.paintColor1 = paintColor1;
        invalidate();
    }

    public void setPaintColor2(int paintColor2) {
        this.paintColor2 = paintColor2;
        invalidate();
    }

    public void saledProgress(int progress, boolean is) {
        offsetAngle = (progress * 360) / 100;
        if (is) {
            progressAngle = 0;
        } else {
            progressAngle = offsetAngle;
        }

        this.progress = 0;

        if (progressAngle <= offsetAngle) {
            invalidate();
            handler.sendEmptyMessageDelayed(1, 50);
        }
    }

    private Handler handler = new Handler() {
        @Override
        public void handleMessage(@NonNull Message msg) {
            super.handleMessage(msg);
            progressAngle = progressAngle + STEP_ANGLE;
            if (progressAngle <= offsetAngle) {
                invalidate();
                handler.sendEmptyMessageDelayed(1, 50);
            }
        }
    };

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);

        if (widthMode == MeasureSpec.AT_MOST){
            progressViewWidth=50;
        }else {
            int size = MeasureSpec.getSize(widthMeasureSpec);

            if (size>=200){
                progressViewWidth=200;
            }else {
                progressViewWidth=size;
            }
        }

        if (heightMode == MeasureSpec.AT_MOST){
            progressViewHeight=50;
        }else {
            int size = MeasureSpec.getSize(heightMeasureSpec);

            if (size>=200){
                progressViewHeight=200;
            }else {
                progressViewHeight=size;
            }
        }

        setMeasuredDimension(progressViewWidth,progressViewHeight);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        progressViewWidth = getMeasuredWidth();
        progressViewHeight = getMeasuredHeight();

        int centerX = progressViewWidth / 2;
        int centerY = progressViewHeight / 2;

        int radius = (progressViewWidth < progressViewHeight ? progressViewWidth / 2 : progressViewHeight / 2) - CIRCLE_MARGIN;

        paint.setColor(paintColor1);
        paint.setAntiAlias(true);
        paint.setStrokeWidth(paintSize);
        paint.setStyle(Paint.Style.STROKE);
        canvas.drawCircle(centerX, centerY, radius, paint);

        RectF rectF = new RectF(progressViewWidth / 2 - radius, progressViewHeight / 2 - radius, progressViewWidth / 2 + radius, progressViewHeight / 2 + radius);
        paint.setColor(paintColor2);
        canvas.drawArc(rectF, START_ANGLE, progressAngle, false, paint);

        Rect rect = new Rect();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(textColor);
        paint.setTextSize(textSize);
        paint.setStrokeWidth(2);
        String content = (progressAngle * 100) / 360 + "%";
        paint.getTextBounds(content, 0, content.length(), rect);
        canvas.drawText(content, progressViewWidth / 2 - rect.width() / 2, progressViewHeight / 2 + rect.height() / 2, paint);
    }
}

File: frame_library/src/main/java/com/example/frame_library/view/ProgressView.kt
package com.example.frame_library.view

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.view.animation.Animation
import android.widget.TextView
import com.example.frame_library.R

class ProgressView(context: Context?, attrs: AttributeSet?, defStyleAttr: Int):View(context, attrs, defStyleAttr) {

    private val paint:Paint= Paint()
    private var textcolor:Int?=null
    private var Star:Boolean?=null
    private var progress:Int?=null
    private var animator:ValueAnimator?=null
    private var color:Int=Color.RED
    var textsize:Float=80f
    var size:Float=50f
    constructor(context: Context?) : this(context,null)
    constructor(context: Context?, attrs: AttributeSet?) : this(context, attrs,0)

    init {
        var obtainStyledAttributes =
            context!!.obtainStyledAttributes(attrs, R.styleable.ProgressView)
        textcolor=obtainStyledAttributes.getColor(R.styleable.ProgressView_textColor,Color.BLACK)
        Star=obtainStyledAttributes.getBoolean(R.styleable.ProgressView_star,true)
        progress=obtainStyledAttributes.getInt(R.styleable.ProgressView_progress,30)
        color=obtainStyledAttributes.getColor(R.styleable.ProgressView_color,Color.RED)
        textsize=obtainStyledAttributes.getInt(R.styleable.ProgressView_textSize,80).toFloat()
        size=obtainStyledAttributes.getInt(R.styleable.ProgressView_size,50).toFloat()
        obtainStyledAttributes.recycle()
    }

    /**
     * @param progress 圆的进度
     */
    fun setProgress(progress:Int){
        this.progress=progress;
        invalidate()
    }

    fun setStar(b:Boolean){
        this.Star=b
        invalidate()
    }

    /***
     * 加载动画
     */
     fun Animation(){
        if (Star!!) {
            animator = ValueAnimator.ofFloat(0f, progress!!.toFloat())
            animator!!.addUpdateListener {
                Star=false
                var t = it.animatedValue as Float

                this.progress=t.toInt()
                invalidate()
            }
            animator!!.duration=3*1000
            animator!!.start()
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        var wMode = MeasureSpec.getMode(widthMeasureSpec)
        var hMode = MeasureSpec.getMode(heightMeasureSpec)
        var wSize=if (wMode==MeasureSpec.AT_MOST){
            500
        }else{
            MeasureSpec.getSize(widthMeasureSpec)
        }
        var hSize=if (hMode==MeasureSpec.AT_MOST){
            500
        }else{
            MeasureSpec.getSize(heightMeasureSpec)
        }
        setMeasuredDimension(wSize,hSize)
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        paint.setColor(Color.GRAY)
        paint.strokeWidth=size
        paint.style=Paint.Style.STROKE
        paint.isAntiAlias=true

        var mWidth = measuredWidth
        var mHeight = measuredHeight

        var viewY = mHeight / 2
        var viewX = mWidth / 2
        var radius=if (mWidth>mHeight){mHeight/2}else{mWidth/2}-paint.strokeWidth/2

        canvas!!.drawCircle(viewX.toFloat(),viewY.toFloat(),radius,paint)

        paint.setColor(color)

        canvas!!.drawArc(viewX-radius,viewY-radius,viewX+radius,viewY+radius,0f,(progress!!*3.6).toFloat(),false,paint)

        paint.color= textcolor!!
        paint.textSize=textsize
        paint.style=Paint.Style.FILL

        var rect = Rect()
        var text = "${progress}%"
        paint.getTextBounds(text,0,text.length,rect)
        canvas.drawText(text,(viewX-rect.width()/2).toFloat(),(viewY+rect.height()/2).toFloat(),paint)

        Animation()
    }

    fun destroy(){
        animator!!.cancel();
        animator=null
    }
}